src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -102,13 +111,228 @@ private static async Task<VersionStamp> GetDocumentSemanticVersionAsync(Document
+                // The range candidate becomes the best range either if:
+                //     1) It is the first candidate we are analyzing
+                //     2) The # of overlapping lines exceeds the current best candidate, *and* the current candidate
+                //     is not in the middle of the requested range. (We avoid the latter since it complicates later
+                //     computations.)
+                var numOverlappingLines = overlap.End.Line - overlap.Start.Line;
+                if (currentBestMatch is null ||
+                    ((currentBestMatch.Value.NumOverlappingLines < numOverlappingLines) && (rangeCandidate.Start <= range.Start || rangeCandidate.End >= range.End)))
+                {
+                    currentBestMatch = (rangeCandidate, numOverlappingLines);
+                }
+            }
+
+            // We couldn't find a best range match. Don't use any cache values and compute tokens normally.
+            if (currentBestMatch is null)
+            {
+                return null;
+            }
+
+            if (!documentCache.TryGetValue(currentBestMatch.Value.Range, out cachedTokens))
+            {
+                logger.LogError("Could not find the expected cached tokens.");
+                return null;
+            }
+
+            // 3) We've found the best range match. There are now two paths to consider depending on whether
+            // we need to populate the start or end of the range.
+            var bestMatch = currentBestMatch.Value;
+
+            // 3a) The latter part of the range needs to be computed (i.e. the beginning is already cached).
+            if (bestMatch.Range.Start <= range.Start)
+            {
+                var tokens = await ComputeRequiredRange(computeStart: false, cachedTokens);
+                if (tokens is null)
+                {
+                    return null;
+                }
+
+                return tokens;
+            }
+            // 3b) The beginning of the range needs to be computed (i.e. the end is already cached).
+            else if (bestMatch.Range.End >= range.End)
+            {
+                var tokens = await ComputeRequiredRange(computeStart: true, cachedTokens);
+                if (tokens is null)
+                {
+                    return null;
+                }
+
+                return tokens;
+            }
+            else
+            {
+                // We shouldn't ever reach this point, so if we do we'll log an error and return.
+                logger.LogError("Caching error in DefaultRazorSemanticTokensInfoService.");
+                return null;
+            }
+
+            async Task<SemanticTokens?> ComputeRequiredRange(bool computeStart, SemanticTokens cachedTokens)
+            {
+                // We want to reduce the requested range to the part we don't already have cached.
+                var modifiedRange = computeStart
+                    ? new Range(range.Start, bestMatch.Range.Start)
```

Comment:
I thinkkkk `bestMatch.Range.Start` will be < `range.Start`, same reversal on the next line
```suggestion
                    ? new Range(bestMatch.Range.Start, range.Start)
```